source/sas.cpp

```diff
@@ -756,40 +758,89 @@ SAS::Timestamp SAS::Marker::get_timestamp() const
+  va_end(args);
+}
 
-  switch (level) {
-    case LOG_LEVEL_ERROR:   level_str = "ERROR"; break;
-    case LOG_LEVEL_WARNING: level_str = "WARNING"; break;
-    case LOG_LEVEL_STATUS:  level_str = "STATUS"; break;
-    case LOG_LEVEL_INFO:    level_str = "INFO"; break;
-    case LOG_LEVEL_VERBOSE: level_str = "VERBOSE"; break;
-    case LOG_LEVEL_DEBUG:   level_str = "DEBUG"; break;
-    default:                level_str = "UNKNOWN"; break;
+
+// Write the internally raised log to a char array, then pass
+// this to the common logging callback to log.
+void SAS::_sasclient_log_callback(log_level_t level,
+                                  const char *module,
+                                  int line_number,
+                                  const char *fmt,
+                                  va_list args)
+{
+  const char* log_level[] = {"Error", "Warning", "Status", "Info", "Verbose", "Debug"};
+  char logline[MAX_LOGLINE];
+  int written = 0;
+
+  // Strip the directory from the file location.
+  const char* mod = strrchr(module, '/');
+  module = (mod != NULL) ? mod + 1 : module;
+
+  if (line_number)
+  {
+    written = snprintf(logline, MAX_LOGLINE - 2, "CHECK %s %s:%d: ", log_level[level], module, line_number);
+  }
+  else
+  {
+    written = snprintf(logline, MAX_LOGLINE - 2, "%s %s: ", log_level[level], module);
   }
 
-  printf("%s %s:%d: ", level_str, module, line_number);
-  vprintf(fmt, args);
-  printf("\n");
-  fflush(stdout);
+  // snprintf and vsnprintf return the bytes that would have been
+  // written if their second argument was large enough, so we need to
+  // reduce the size of written to compensate if it is too large.
+  written = std::min(written, MAX_LOGLINE - 2);
 
-  va_end(args);
-}
+  int bytes_available = MAX_LOGLINE - written - 2;
+  written += vsnprintf(logline + written, bytes_available, fmt, args);
 
+  if (written > (MAX_LOGLINE - 2))
+  {
+    written = MAX_LOGLINE - 2;
+  }
 
-void SAS::discard_logs(log_level_t level,
-                       const char *module,
-                       int line_number,
-                       const char *fmt,
-                       ...)
-{
+  // Convert log_level_t to sasclient_log_level_t, which is the type expected by the common
+  // logging function.
+  sasclient_log_level_t sas_level;
+  switch (level) {
+    case LOG_LEVEL_ERROR:
+      sas_level = SASCLIENT_LOG_ERROR;
+      break;
+    case LOG_LEVEL_WARNING:
+      sas_level = SASCLIENT_LOG_WARNING;
+      break;
+    case LOG_LEVEL_STATUS:
+      sas_level = SASCLIENT_LOG_INFO;
+      break;
+    case LOG_LEVEL_INFO:
+      sas_level = SASCLIENT_LOG_INFO;
+      break;
+    case LOG_LEVEL_VERBOSE:
+      sas_level = SASCLIENT_LOG_DEBUG;
+      break;
+    case LOG_LEVEL_DEBUG:
+      sas_level = SASCLIENT_LOG_DEBUG;
+      break;
+    default:
+      SAS_LOG_WARNING("Unknown SAS log level %d, treating as error level", level);
+      sas_level = SASCLIENT_LOG_ERROR;
+    }
```

Comment:
Do you need to do this conversion? Why not eliminate the internal `LOG_LEVEL_*` variables entirely and just use the `SASCLIENT_LOG_*` constants throughout?

Reply:
Because when this gets passed to the CPP-common log callback we need to check whether or not to write the log against the global logging level.

Reply:
What I'm saying is with this PR we have three logging levels (using error logs as an example):

* `LOG::ERROR_LEVEL` defined in log.h
* `SASCLIENT_LOG_ERROR` which is exposed on the sas-client interface. 
* `SAS::LOG_LEVEL_ERROR` which is used in the `SAS_LOG_ERROR` macro. 

Why not get rid of the last set, and have `SAS_LOG_ERROR` just use the values defined on the interface?

Reply:
I see, I agree.